### ISSUE TITLE
Users are receiving repeat e-mails every 6 hours

### ISSUE THREAD

**Issue by bowlofeggs** [fedora-infra/bodhi#1009]
@kparal reported than [an Epiphany update](https://bodhi.fedoraproject.org/updates/FEDORA-2016-f4b5897686) has been e-mailing him every 6 hours with the push to stable message, like this:

```
This update has reached 11 days in testing and can be pushed to stable now if the maintainer wishes
```

He received one at 2016-10-04T08:00, 2016-10-04T14:00, and 2016-10-04T20:00, for example.

We should ensure that these messages only get sent once.

**Comment by kparal** [comment]
The interesting thing is that I received those emails (about 5 of them I think), but no bodhi comment was added. I'd expect a bodhi comment to be added (as is usually is in this case) and therefore receiving an email.

Also, it seems that every commenter of this update received such emails, according to the response in that update.

**Comment by bowlofeggs** [comment]
That is interesting to note. Perhaps it indicates that there was some error and a transaction was rolled back, which could lead to Bodhi not knowing it had already sent the e-mails. I'm not familiar with that area of the code yet, so that's just a wild guess.

**Comment by richardfearn** [comment]
I think the reminder comments and/or emails haven't been working for a couple of weeks. I realised recently that some of my updates (e.g. https://bodhi.fedoraproject.org/updates/FEDORA-2016-3a5fa75880) haven't had the reminder comment added, nor have I received the emails.

**Comment by lubomir** [comment]
For another data point: I gave karma to [an update](https://bodhi.fedoraproject.org/updates/FEDORA-2016-4da6039b17) yesterday at 15:17, and starting at 20:00 I started receiving the notifications (one every 6 hours). (I doesn't really bother me, I'm good at ignoring e-mails.)

**Comment by bowlofeggs** [comment]
I believe I know the cause of both getting too many e-mails, and getting none. The [code that generates these e-mails](https://github.com/fedora-infra/bodhi/blob/2.2.4/bodhi/server/scripts/approve_testing.py#L52-L73) looks like this:

```
    with transaction.manager:
        testing = db.query(Update).filter_by(status=UpdateStatus.testing,
                                             request=None)
        for update in testing:
            # If this release does not have any testing requirements, skip it
            if not update.release.mandatory_days_in_testing:
                print('%s doesn\'t have mandatory days in testing' % update.release.name)
                continue
            # If this has already met testing requirements, skip it
            if update.met_testing_requirements:
                continue
            if update.meets_testing_requirements:
                print('%s now meets testing requirements' % update.title)
                text = config.get('testing_approval_msg') % update.days_in_testing
                update.comment(db, text, author='bodhi')

            # Approval message when testing based on karma threshold
            if update.stable_karma not in (0, None) and update.karma >= update.stable_karma \
                    and not update.autokarma:
                print('%s now reaches stable karma threshold' % update.title)
                text = config.get('testing_approval_msg_based_on_karma')
                update.comment(db, text, author='bodhi')
```

I think there are two contributing factors at play here. The first is that the approve testing executable is making a single transaction for all updates, meaning that if anything within them is an error all changes will be rolled back. The second is that 2.2.0 introduced a new setting, called `testing_approval_msg_based_on_karma` and that setting is not present in production. Therefore, what I think is happening is that:
1. Every 6 hours this job runs,
2. A few e-mails get sent out as various updates meet the requirements,
3. Until an update with enough karma is reached in the query that wants to use that setting,
4. Since the setting doesn't exist, this causes a traceback which causes the transaction to roll back, which explains why we don't see the comments on the updates from Bodhi despite the e-mails having already gone out (you can't roll back sent e-mails ☺).

I think we should fix this two ways:
1. Most obviously, we need to add that setting to production. This will require a Freeze Break Request, since Fedora is currently in a beta freeze. I'll file one today.
2. I believe we should be committing our transaction as we go through the for loop, so that we at least save the changes we make to the database as we go.

While I was in there, I noticed the call to [`met_testing_requirements()`](https://github.com/fedora-infra/bodhi/blob/develop/bodhi/server/models/models.py#L1791-L1812), which looks like this:

```
        min_num_days = self.release.mandatory_days_in_testing
        if min_num_days:
            if not self.meets_testing_requirements:
                return False
        else:
            return True
        for comment in self.comments:
            if comment.user.name == 'bodhi' and \
               comment.text.startswith('This update has reached') and \
               'days in testing and can be pushed to stable now if the ' \
               'maintainer wishes' in comment.text:
                return True
        return False
```

I don't like that this code is using this startwith/endswith stuff, because which message gets put on the comment is actually a setting (see production/staging/development.ini files). Therefore, the code as it is is depending on the setting being set a particular way. I've filed https://github.com/fedora-infra/bodhi/issues/1014 about that.

**Comment by bowlofeggs** [comment]
I will use this issue to track fixing the transaction handling, so problems that happen during the run don't roll back any changes that were made (sending confusing ghost e-mails).

I will use [Richard Fearn's infrastructure ticket](https://pagure.io/fedora-infrastructure/issue/5515) to track fixing the missing setting in production.

**Comment by bowlofeggs** [comment]
The production configuration changes are in place on production now. I will keep this ticket open to track fixing the transactions to be more granular. Thanks for the report and the patience everyone!

**Comment by bowlofeggs** [comment]
Upon further reflection, I believe that fixing the production settings will solve only part of the issue. There is still a problem that can only be fixed by editing the code. `met_testing_requirements()` as cited above only checks for the `testing_approval_msg` having been added to the update, which happens when the update has been in testing for the required amount of time. Unfortunately, that means that updates that hit the karma threshold before the required number of days are going to get the "This update has reached the stable karma threshold and can be pushed to stable now if the maintainer wishes." message every six hours until the required number of days is reached, or until the maintainer pushes the update to stable.

One way we can prevent that from happening is if we fix `met_testing_requirements()` to check for either message instead of only checking for the one about length of time. I still think we can come up with a better check than looking for a comment with particular text in it, but for now we might want a quick fix, such as making `met_testing_requirements()` look like this:

```
            if comment.user.name == 'bodhi' and \
               comment.text.startswith('This update has reached') and \
               'can be pushed to stable now if the ' \
               'maintainer wishes' in comment.text:
                return True
```

The only difference is that I removed "days in testing and " from the second half of the check. It's still not a good way to do it, but I've got another open ticket to fix the way this works and right now a quick fix might be what we want.

**Comment by bowlofeggs** [comment]
I decided to file https://github.com/fedora-infra/bodhi/issues/1016 to track fixing the transaction, so that this ticket can focus solely on fixing the past-comment-detection.

**Comment by richardfearn** [comment]
Thanks for looking into this so quickly! I've started getting the reminder emails again now. I did spot another issue and I'll file another issue about that.

**Comment by bowlofeggs** [comment]
This should get fixed in the upcoming 2.3.0 release. As of now all the release blockers that I know of are merged, so I will begin the release process for 2.3.0 soon.

**Comment by bowlofeggs** [comment]
According to #1116, this issue is not solved.

**Comment by bowlofeggs** [comment]
I've spent some time looking at the code and the way it tests for these messages (which isn't nice), but I can't tell why https://bodhi.fedoraproject.org/updates/FEDORA-2016-f5e78ec298 got all those messages. I deployed bodhi-2.3.3 to production last week (I believe on Wednesday, Nov 30).  Can anyone confirm whether these repeat e-mails have happened since then?

**Comment by bowlofeggs** [comment]
I poked around at some updates that don't use autokarma, and I found that the kernel is still having this problem:

https://bodhi.fedoraproject.org/updates/FEDORA-2016-9c17cb9648

**Comment by bowlofeggs** [comment]
Ah, I believe I know what is happening and why I mistakenly thought it was fixed before. I believe the problem was solved for all updates except for non-autokarma critical path updates, which have slightly different rules. Both examples that we know about are critical path!

non-autokarma updates get [this check](https://github.com/fedora-infra/bodhi/blob/2.3.3/bodhi/server/scripts/approve_testing.py#L70-L74) to determine whether Bodhi should write that comment or not:

```
            if update.stable_karma not in (0, None) and update.karma >= update.stable_karma \
                    and not update.autokarma:
                print('%s now reaches stable karma threshold' % update.title)
                text = config.get('testing_approval_msg_based_on_karma')
                update.comment(db, text, author='bodhi')
```

Critical path updates cannot be pushed until they've reached ```critpath.stable_after_days_without_negative_karma```, which is defined in the config file on production as 14 days. Because of this, ```Update.meets_testing_requirements``` will return False for the kernel and rpm Updates, regardless of how much karma they have. There is a check in the loop to test if the comment has already been added that calls ```Update.met_testing_requirements```. This in turn calls ```meets_testing_requirements``` which returns False, so it returns False too. Then the check to add the comment only looks at the karma level and not the critpath rules, so it writes the comment every time.

I suspect we can just add an ```and meets_testing_requirements``` to the bit that writes the comment to fix this correctly. As I noted before, this code is way too complicated for what it does, and the comment detection is very fragile too. We really need to rewrite this area of the code to be clean so fixing problems like this isn't so difficult. I don't have time to do that now, but I made a note of it so I find some time to do it later.
